Azure OpenAI: TypeSpec updates for 2023-12-01-preview

trrwilson opened the pull request:
This PR transfers to TypeSpec the changes made in the `2023-12-01-preview` service API version for Azure OpenAI:
https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview

trrwilson committed:
baseline: add new label with no changes

trrwilson committed:
reapplication of changes from PR repository

trrwilson committed:
incremental: chat inputs, including tools

trrwilson committed:
incremental, in-progress: chat message rework

trrwilson committed:
incremental: chat completions updated (undoc'd/unverified)

trrwilson committed:
incremental: provisional completeness for chat

trrwilson committed:
Incremental: RAI updates

kristapratico commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_chat_extensions.tsp:
```diff
@@ -1,7 +1,11 @@
-
 import "./chat_completions.tsp";
-import "./azure_cognitive_search_options.tsp";
+import "./chat_messages.tsp";
+import "./extensions_onyourdata_azure_cognitive_search_options.tsp";
+import "./extensions_onyourdata_azure_machinelearning_index_options.tsp";
+import "./extensions_onyourdata_azure_cosmosdb_options.tsp";
+import "./extensions_onyourdata_elasticsearch_options.tsp";
+import "./extensions_onyourdata_pinecone_options.tsp";
```
It looks like the additional data source options got removed for the 2023-12-01-preview swagger that merged to the public repo. Should we leave them out of the typespec as well?

trrwilson replied:
I believe they're currently *missing* from the swagger file. We believe they'll still end up implicitly supported since routing will succeed, but we need to follow up with the OYD team to ensure we get the right consistency and APIM support.

trrwilson committed:
Merge branch 'main' into user/travisw/aoai-2023-12-01-preview-typespec

kristapratico commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,165 @@
+  filtered: boolean;
+
+  @doc("A value indicating whether detection occurred, irrespective of severity or whether the content was filtered.")
+  detected: string;
```
I think this should be a boolean. At least, that's what I'm seeing returned from the API:

```json
{
    "error": {
        "message": "The response was filtered due to the prompt triggering Azure OpenAI's content management policy. Please modify your prompt and retry. To learn more about our content filtering policies please read our documentation: https://go.microsoft.com/fwlink/?linkid=2198766",
        "type": null,
        "param": "prompt",
        "code": "content_filter",
        "status": 400,
        "innererror": {
            "code": "ResponsibleAIPolicyViolation",
            "content_filter_result": {
                "custom_blocklists": [
                    {
                        "filtered": true,
                        "id": "CustomBlockList8"
                    }
                ],
                "hate": {
                    "filtered": false,
                    "severity": "safe"
                },
                "jailbreak": {
                    "detected": false,
                    "filtered": false
                },
                "profanity": {
                    "detected": false,
                    "filtered": false
                },
                "self_harm": {
                    "filtered": false,
                    "severity": "safe"
                },
                "sexual": {
                    "filtered": false,
                    "severity": "safe"
                },
                "violence": {
                    "filtered": false,
                    "severity": "safe"
                }
            }
        }
    }
}
```

trrwilson replied:
Thanks for confirming! I suspected that, too--I was trying to faithfully copy from the One True Swagger, but that got a "hmm" as I read it. I'll fix it in both places. @bexxx FYI that the OpenAPI docs have the type of `detected` misrepresented.

kristapratico commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,165 @@
+
+@added(ServiceApiVersions.v2023_06_01_Preview)
+@doc("Information about content filtering evaluated against generated model output.")
+model ContentFilterResultsForChoice {
```
Should this model also have the protected_material_text and protected_material_code? 

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.json#L1137

trrwilson replied:
Absolutely should. Will add that now along with the `detected` fix.

trrwilson committed:
RAI, data-plane, tsp compilation fixes

trrwilson committed:
minor updates after starting .NET integration

trrwilson committed:
small fix to structured text items for vision

trrwilson committed:
small but important function fix

trrwilson committed:
trial and error: OYD apiKey->key

trrwilson committed:
minor edit

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/tspconfig.yaml:
```diff
@@ -20,14 +20,15 @@ options:
     generate-tests: false
     custom-types-subpackage: "implementation.models"
-    custom-types: "FunctionCallPreset"
+    custom-types: "FunctionCallModelBase,FunctionCallPreset,FunctionCallPresetFunctionCallModel,FunctionNameFunctionCallModel"
```
I think you need to merge `main`. There were recently removed due to the `access/usage` change.

trrwilson committed:
java emitter config merge

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,175 @@
+
+  @doc("A value indicating whether or not the content has been filtered.")
+  @projectedName("json", "filtered")
+  filtered: boolean;
```
Do you think we may want to use `contentFilterResultBase` (notice singular `result`) verbatim from what there is in the service? It would be a model that looks like:

```ts
model ContentFilterResultBase {
  filtered: boolean;
}
```

Then wherever (well, not wherever, but I will mark the places where the swagger seems to use a spread operator) we have `filtered` we use instead `...ContentFilterResultBase`

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,175 @@
+@doc("Represents the outcome of a detection operation performed by content filtering.")
+model ContentFilterDetectionResult {
+  @doc("A value indicating whether or not the content has been filtered.")
+  filtered: boolean;
```
This could be `...ContentFilterResultBase`

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,175 @@
+  id: string;
+
+  @doc("A value indicating whether or not the content has been filtered.")
+  filtered: boolean;
```
This could be `...ContentFilterResultBase`

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp:
```diff
@@ -204,17 +168,46 @@ model ChatCompletionsOptions {
+
+  @added(ServiceApiVersions.v2023_12_01_Preview)
+  @doc("If provided, the configuration options for available Azure OpenAI chat enhancements.")
+  enhancements?: AzureChatEnhancementConfiguration;
```
I don't want to add to the pile here, but if we are incurring into already plenty of API breaking changes we may want to go all in and describe `AzureChatCompletionsOptions` in a way that composes the fields closer to what is done in the swagger:

```ts
model ChatCompletionsOptionsBase {
  // the usual fields but we move `dataSources` to an azure dedicated model for better compo
  // We also take all the fields that will show up in the model below
  // ...
}

model ChatCompletionsOptions {
  ...ChatComplettionsOptionsBase;
  messages: ChatMessages[];
  n: integer;
  seed: integer;
  response_format: ChatCompletionsResponseFormat;
  tools: ChatCompletionsToolDefinition[];
  tool_choice: ChatCompletionsToolSelection;
  functions: FunctionDefinition[];
  function_call: FunctionCallConfig;
}

model AzureChatCompletionsOptions {
  ...ChatCompletionsOptionsBase;
  messages: ChatMessages[]; 
  dataSources?: AzureChatExtensionConfiguration[];
  enhancements?: AzureChatExtensionEnhancements;
}
```

jpalvarezl commented on specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2023-05-15/generated.json:
```diff
@@ -345,7 +345,7 @@
           "description": "The collection of context messages associated with this chat completions request.\nTypical usage begins with a chat message for the System role that provides instructions for\nthe behavior of the assistant, followed by alternating messages between the User and\nAssistant roles.",
           "items": {
-            "$ref": "#/definitions/ChatMessage"
+            "$ref": "#/definitions/ChatRequestMessage"
```
The rename of the model appears to be retroactive. But only for `2023-05-01-stable`. Maybe this just needs reverting.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/images/images.tsp:
```diff
@@ -10,80 +10,170 @@ using TypeSpec.Versioning;
+  Supported by both dall-e-2 and dall-e-3 models.
+  """)
+  @added(ServiceApiVersions.v2023_12_01_Preview)
+  size1024x1024: "1024x1024",
```
nit: maybe we should mention this is the default for azure services in the doc.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/images/images.tsp:
```diff
@@ -10,80 +10,170 @@ using TypeSpec.Versioning;
 
   @doc("Image generation response items should provide image data as a base64-encoded string.")
-  Base64: "b64_json",
+  base64: "b64_json",
```
I prefer this the way it is, but should we really drop `_json`? This rings a similar bell as keeping `n` as "n" since people are already used to it.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/images/images.tsp:
```diff
@@ -10,80 +10,170 @@ using TypeSpec.Versioning;
 @added(ServiceApiVersions.v2023_06_01_Preview)
 model ImageGenerationOptions {
+  @doc("The model to use for image generation.")
+  `model`?: string;
```
not entirely sure how this was working before, but shouldn't we annotate this `added` ?

trrwilson replied:
It was technically always supported, just optional and had a single value.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/images/images.tsp:
```diff
@@ -10,80 +10,170 @@ using TypeSpec.Versioning;
+  @doc("The complete data for an image, represented as a base64-encoded string.")
   @projectedName("json", "b64_json")
-  base64Data: string;
+  base64Data?: string;
```
Not entirely sure if worth the breaking change, but just in case, we could match the field name here to what we have for `ImageGenerationResponseFormat` and just go with `base64`

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cognitive_search_options.tsp:
```diff
@@ -71,35 +81,33 @@ model AzureCognitiveSearchChatExtensionConfiguration {
   embeddingEndpoint?: url;
 
+  @added(ServiceApiVersions.v2023_08_01_Preview)
+  @removed(ServiceApiVersions.v2023_12_01_Preview)
```
We have both added and removed. Could be a pasting error, just FYI.

trrwilson replied:
The `@added` is redundant with the containing model having the same label; I'll remove those where I see them. Doesn't hurt anything but also very unnecessary.

trrwilson replied:
Actually, I take that back -- I'm not sure if it's a compiler quirk, but if I remove the `@added` I get:

> 'Azure.OpenAI.AzureCognitiveSearchChatExtensionConfiguration' was added on version 'v2023_08_01_Preview' but contains type 'Azure.OpenAI.AzureCognitiveSearchChatExtensionConfiguration.embeddingKey' added in version 'v2022_12_01'.TypeSpec(@typespec/versioning/incompatible-versioned-reference)

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_chat_extensions.tsp:
```diff
@@ -30,8 +34,25 @@ using TypeSpec.Versioning;
 enum AzureChatExtensionType {
   @doc("Represents the use of Azure Cognitive Search as an Azure OpenAI chat extension.")
   azureCognitiveSearch: "AzureCognitiveSearch",
```
Just an ack: the variant consuming this discriminator value has been correctly updated to inherit from `AzureChatExtensionConfiguration`

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_chat_extensions.tsp:
```diff
@@ -67,5 +88,86 @@ model AzureChatExtensionsMessageContext {
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A representation of the available options for the Azure OpenAI grounding enhancement.")
+model AzureChatGroundingEnhancementConfiguration {
+  @doc("Specifies whether the enhancement is enabled.")
```
The documentation in the `yaml` is the following: `Request object to specify if grounding enhancement is needed.`. I can't quite put my finger on it, but it sounds like there is some nuance lost (or added) with the different wording. If possible, would be cool to document briefly what is "grounding" in this context.

trrwilson replied:
Enhancements needs a major pass in general. The naming doesn't seem to reflect the members, the documentation doesn't seem to reflect the data, etc.. We'll want to follow up on that in depth and add whatever detail we can to the docs as we test, exercise, and support it.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_chat_extensions.tsp:
```diff
@@ -67,5 +88,86 @@ model AzureChatExtensionsMessageContext {
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A representation of the available options for the Azure OpenAI optical character recognition (OCR) enhancement.")
+model AzureChatOCREnhancementConfiguration {
+  @doc("Specifies whether the enhancement is enabled.")
```
I think the documentation here suffers from the same as the above. Would be great to provide more details here, but it's a total nit pick too (in the current scheme of things 😄)

trrwilson replied:
Not a nit, it's important. Just also something I think we need to follow up on in parallel to getting (insufficiently documented and likely suboptimally structured) support available.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp:
```diff
@@ -3,75 +3,37 @@ import "@typespec/http";
-  @added(ServiceApiVersions.v2023_08_01_Preview)
-  @projectedName("csharp", "AzureExtensionsContext")
-  context?: AzureChatExtensionsMessageContext;
+  jsonObject: "json_object",
```
We might be missing a `@added(ServiceApiVersions.v2023_12_01_Preview)` here.

trrwilson replied:
The enum `ChatCompletionsResponseFormat` has it.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp:
```diff
@@ -3,75 +3,37 @@ import "@typespec/http";
 Completions support a wide variety of tasks and generate text that continues from or "completes"
 provided prompt data.
 """)
+@added(ServiceApiVersions.v2023_05_15)
```
Oh, I see now why we had an name update in the `data-plane/stable` folder for that service release date. Is this on purpose or maybe a pasting mistaking?

trrwilson replied:
That's now our first supported label with chat completions support and adding newly-versioned children items into the unversioned parent was causing issues. I believe the 5-15 backdating is accurate.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp:
```diff
@@ -3,75 +3,37 @@ import "@typespec/http";
-}
-
-@doc("A single, role-attributed message within a chat completion interaction.")
-model ChatMessage {
```
neither `ChatRole` nor `ChatMessage` need to be annotated with `@removed`? I understand that these are not retired models from the service but were just refactored. Is that why the annotation is not needed in this particular type of "removal"?

trrwilson replied:
`ChatRole` was moved into `common.tsp`, while `ChatMessage` was structurally replaced. There are "renamed" decorators and whatnot, but I'm not sure how we'd represent the decomposition replacement that happened with `ChatMessage` becoming a whole family of related things.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
```
super nit: In the `streaming|non-streaming` model separation PR I extracted `alias NullableFloat = float32 | null;` into `common.tsp` . Seeing this, made me think we should probably do the same. Maybe extract this into a `.tsp` file of its own with all these nullable types.

jpalvarezl replied:
Messing a bit in the TSP playground, something like:

```ts
alias NullableField<T> = T | null;

model Dog {
  name: NullableField<string>;
}
```

Yields the following OpenAPIv3 definition:

```yaml
openapi: 3.0.0
info:
  title: (title)
  version: 0000-00-00
tags: []
paths: {}
components:
  schemas:
    Dog:
      type: object
      required:
        - name
      properties:
        name:
          type: string
          nullable: true
```

Totally aware that now is not the time for such housekeeping though 😄

trrwilson replied:
Yeah, and I think the generics mess with emitters in funny ways. Let's do this all at once (ideally together with the change you're working on).

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+@discriminator("type")
+model ChatMessageContentItem {
+  @doc("The discriminated object type.")
+  type: string;
```
The type of of `type` (gosh) seems to be refered as `chatCompletionRequestMessageContentPart` in the yaml. Maybe we should include the enum as well, to have the representation match more strictly and not leaving it so open eneded? Not entirely sure what are the ramifications of using `@discriminator` with a `string` instead of an `enum`

trrwilson replied:
Per the comment a bit later, I'd like to bring up the holistic topic of whether we should strongly type all discriminators as named enums. Then we can switch everything all at once.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  high: "high",
+}
+
+@added(ServiceApiVersions.v2023_05_15)
```
I am not sure this is entirely correct. Did we support this schema from that version? Could be a pasting mistake, just checking :)

trrwilson replied:
Not a mistake, but you're correct that it's also not right.

I'm not sure how to represent this entirely correctly in the TypeSpec such that it won't break the emitters. At the top of the file, we have the `alias ChatMessageContent = string | ChatMessageContentItem[]` thing -- in reality, that second part of that union is what needs to be versioned. There are ways to change types with the versioning library but I'm not sure how to apply it to a *union* like this. Worth following up on, but a known, ugly compromise for the moment.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  The evaluation quality setting to use, which controls relative prioritization of speed, token consumption, and
+  accuracy.
+  """)
+  detail?: ChatMessageImageDetailLevel;
```
I believe that the default is `auto`

```suggestion
  detail?: ChatMessageImageDetailLevel = ChatMessageImageDetailLevel.auto;
```

trrwilson replied:
Deferring this to a holistic defaults pass -- we have a ton that we could add.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  content: string;
+
+  @doc("An optional name for the participant.")
+  name?: string;
```
This appears to be a pasting mistake. I don't see any `name` optional property in the yaml file. (looking at `chatCompletionRequestMessageSystem`, FYI)

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  content: ChatMessageContent;
+
+  @doc("An optional name for the participant.")
+  name?: string;
```
Also here, I don't see `name` in the yaml file.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  content: string | null;
+
+  @doc("An optional name for the participant.")
+  name?: string;
```
Ok this is probably a non-Azure field. Sorry for all the noise. I will leave the comments just in case.

trrwilson replied:
Right, these all show up in [OpenAI's docs](https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages), though they curiously don't seem to appear [in the matching spec](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5617C9-L5617C9). I know these for sure worked in the past (you could provide a name in the user message and then the assistant would "know your name") so we should likely leave these to begin with and follow up to see if removal was intended.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  The tool calls that must be resolved and have their outputs appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  toolCalls?: ChatCompletionsToolCall[];
```
I think that the yaml uses specifically the concretization `ChatCompletionsFunctionToolCall`. Seems to always be a `chatCompletionMessageToolCall` (yaml file name)

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  The tool calls that must be resolved and have their outputs appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  toolCalls?: ChatCompletionsToolCall[];
```
I think that the yaml uses specifically the concretization `ChatCompletionsFunctionToolCall`. Seems to always be a `chatCompletionMessageToolCall` (yaml file name)

trrwilson replied:
IMO the specs are wrong -- Azure matches [OpenAI's spec](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5833C15-L5833C15) in this regard, with "function" inside of the generic "tool call," but that's at odds with the pattern where the discriminator-named property only shows up in the hard-coded concrete type that has that name. It's "a nice simplification" for now from the REST structure, but if/when a new tool type is introduced, anticipating the correct structure (where the enum gets more than just "function" and the function property has to move to an extended model) avoids a breaking change in generated object models.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+
+  @projectedName("json", "tool_call_id")
+  @doc("The ID of the tool call resolved by the provided content.")
+  toolCallId: string;
```
This [line](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1086) makes me think that `toolCallId` might also be nullable.

trrwilson replied:
[OpenAI spec](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5704C6-L5704C6) has it required, non-nullable.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+@doc("A representation of a chat message as received in a response.")
+model ChatResponseMessage {
+  @doc("The chat role associated with the message.")
+  role: ChatRole;
```
Correct me if I am wrong here, but this seems to **always** be assistant (at least for azure): the `role` model is separate in the response scenario ([line](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1149)) and that enum only contains assistant according to [this](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1161)

trrwilson replied:
Hmm, need to think about this one more:
- [OpenAI's documentation](https://platform.openai.com/docs/api-reference/chat/object#chat/object-choices) doesn't constrain it
- But [OpenAI's spec](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5907C10-L5907C10) does
- A sticky point might be figuring out what to do with the "context" messages for dataSources; are those a "special" kind of request message?

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  toolCalls?: ChatCompletionsToolCall[];
+
+  @added(ServiceApiVersions.v2023_07_01_Preview)
+  @removed(ServiceApiVersions.v2023_12_01_Preview)
```
`function_calls` doesn't appear to be removed [here](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1159)

trrwilson replied:
Thanks; I'm not entirely sure how we should handle "deprecated" effectively, but fully agreed that we shouldn't remove it from our generated spec if it'll accepted by the service.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  If Azure OpenAI chat extensions are configured, this array represents the incremental steps performed by those
+  extensions while processing the chat completions request.
+  """)
+  context?: AzureChatExtensionsMessageContext;
```
I don't see `context` in the definition. But I am suspecting that we are conflating `ChatCompletionMessage` in the extension and non-extension scenario?

trrwilson replied:
Right. We're not separating /chat/completions and /extensions/chat/completions models in the TypeSpec.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/tools.tsp:
```diff
@@ -0,0 +1,97 @@
+  function: FunctionDefinition;
+}
+
+// tool calls: received in responses
```
I think here the service definition becomes really confusing. The model for a tool call for a response seems to be also used in the `ChatCompletionsRequestMessageAssitant`.

jpalvarezl commented on specification/cognitiveservices/OpenAI.Inference/models/completions/tools.tsp:
```diff
@@ -0,0 +1,97 @@
+@doc("""
+The definition information for a chat completions function tool that can call a function in response to a tool call.
+""")
+model ChatCompletionsFunctionToolDefinition
```
I don't think this model is ever used. You can probably delete it, but just to document more or less what I think I see from the yaml:

There are the following structures related to tools:

```ts
// Used in request, variant of `ChatMessage`
model chatCompletionRequestMessageTool {
  tool_call_id: string;
  content: string;
}

//Used in request, variant of `ChatMessage`
model chatCompletionMessageToolCall {
  type: "function";
  function: FunctionCall;
}

model chatCompletionRequestMessageAssistant {
  content: string;
  tool_calls: chatCompletionMessageToolCall[];
}
```

Finally the response calls:

```ts
model ChatResponseMessage {
  // all the other stuff
  tool_calls: chatCompletionMessageToolCall[]; // the same as in the request message for assistant
}
```

There is never a `FunctionDefinition` (but there is a `FunctionCall`) in the context of `tool_calls`

trrwilson replied:
It's used in the request options when specifying the new tools:
```
  tools?: ChatCompletionsToolDefinition[];
```
`ChatCompletionsToolDefinition` is the "abstract" discriminated type, and then `ChatCompletionsFunctionToolDefinition` is a "concrete" derivation of it.

Maps to here: https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools

jpalvarezl commented:
One last totally deferrable nit: grouping everything prefixed `extension_` to an `extension` folder. And any subsequent folders that also could make sense, like `onyourdata` or `oyd` or anything better :)

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp:
```diff
@@ -0,0 +1,202 @@
+  name: string;
+
+  @doc("The output of the function as requested by the function call.")
+  content: string;
```
[OpenAI documents it](https://platform.openai.com/docs/api-reference/chat/create) as a required, non-nullable string; [their specification](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5723C25-L5723C25) meanwhile does agree with it being nullable. I'll side with the spec. Thanks for catching this!

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/tools.tsp:
```diff
@@ -0,0 +1,97 @@
+@added(ServiceApiVersions.v2023_12_01_Preview)
+model ChatCompletionsToolCall {
+  @doc("The object type.")
+  type: string;
```
I'd like to defer this one to the holistic question of "should we use (named) enums for all discriminators?" -- we should apply the yes/no to that everywhere, as we currently have a mix of the two.

trrwilson committed:
some PR feedback + tool_calls role fix

trrwilson committed:
small but important: url is optional in protected_material_code

trrwilson committed:
add missing layer to image_url

trrwilson committed:
Add speculative finish_details for gpt-4-vision-preview

jpalvarezl committed:
Reorging files for extensions

jpalvarezl committed:
Using generics

jpalvarezl committed:
Removed generics approach

jpalvarezl committed:
Fixed various typos and added docs

jpalvarezl committed:
Removed super class parameters

jpalvarezl committed:
Removed LRO entries in client.tsp, since they are removed from spec anyway

jpalvarezl committed:
WIP

jpalvarezl committed:
removed LRO related definitions

deyaaeldeen commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,175 @@
+
+  @doc("Whether a jailbreak attempt was detected in the prompt.")
+  @added(ServiceApiVersions.v2023_12_01_Preview)
+  jailbreak: ContentFilterDetectionResult;
```
Is this always present for it to be required or is it a mistake?

trrwilson replied:
Great catch! I've always observed it in the actual responses thus far, but [per the spec](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.json#L1113) it's not listed as required. I'll update the TypeSpec here to reflect that.

jpalvarezl committed:
Added @client to hide the LRO op

jpalvarezl committed:
Moved visibility modifiers to their own source file

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_authentication.tsp:
```diff
@@ -0,0 +1,89 @@
+  connectionString: string;
+}
+
+@doc("The authentication option using Elasticsearch key and keyId pair.")
```
nit: "keyId" --> "key ID"

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_authentication.tsp:
```diff
@@ -0,0 +1,89 @@
+  type: OnYourDataAuthenticationType;
+}
+
+@doc("The authentication option using API key.")
```
nit: A more complete description might read better for this one and the other derived types, e.g.: "The authentication options for the Azure OpenAI "on your data" extension when using an API key."

trrwilson replied:
Agreed, I'll edit these across the board.

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_authentication.tsp:
```diff
@@ -0,0 +1,89 @@
+@added(ServiceApiVersions.v2023_08_01_Preview)
+model OnYourDataApiKeyAuthenticationOptions
+  extends OnYourDataAuthenticationOptions {
+  @doc("The Authentication type of API key.")
```
nit: "The Authentication type" --> "The authentication type". Applies to the other derived types too.

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_common_options.tsp:
```diff
@@ -0,0 +1,30 @@
+import "@typespec/versioning";
+import "./azure_chat_extensions.tsp";
```
Are these imports circular references? Is this a problem?

- `extensions_onyourdata_common_options.tsp` imports:
- ----> `./azure_chat_extensions.tsp`, which imports:
- --------> `./extensions_onyourdata_azure_cognitive_search_options.tsp` which imports:
- ------------> `./azure_chat_extensions.tsp`, and also `./extensions_onyourdata_common_options.tsp`

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cognitive_search_options.tsp:
```diff
@@ -23,15 +24,20 @@ model AzureCognitiveSearchIndexFieldMappingOptions {
   @doc("The names of index fields that should be treated as content.")
   @projectedName("csharp", "ContentFieldNames")
-  contentFieldNames?: string[];
+  contentFields?: string[];
```
Should this model be renamed to `AzureCognitiveSearchIndexFieldsMappingOptions` (notice "Fields" instead of "Field") for consistency?

trrwilson replied:
I think it's right with "field mapping options." Just like you could eat multiple hot dogs while following "hot dog eating instructions," the infix is about the rules/patterns irrespective of how many instances they're applied over.

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cognitive_search_options.tsp:
```diff
@@ -60,7 +66,11 @@ enum AzureCognitiveSearchQueryType {
+@added(ServiceApiVersions.v2023_08_01_Preview)
+model AzureCognitiveSearchChatExtensionConfiguration
+  extends AzureChatExtensionConfiguration {
+  ...OnYourDataCommonChatExtensionConfiguration;
```
I might be misreading this, but I think all these properties go under a "parameters" property?

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.json#L1945

trrwilson replied:
You're completely correct. This representation was wrong -- the merge of the branch @jpalvarezl has (#26790) addresses this inconsistency.

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_vector_search.tsp:
```diff
@@ -0,0 +1,63 @@
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("Embedding dependency types for vector search.")
+enum OnYourDataEmbeddingDependencyType {
+  @doc("Azure OpenAI Embedding endpoint URL.")
+  endpoint: "Endpoint",
+
+  @doc("Azure OpenAI Embedding deployment name.")
+  deploymentName: "DeploymentName",
```
Are these referring to the endpoint of an Azure OpenAI resource that has an embeddings model deployed and the corresponding deployment name? If so, I would recommend mentioning so in the description.

trrwilson replied:
I'll start a conversation with the OYD team for this and the subsequent questions--agreed that we should clarify.

wangyuantao replied:
Yes. Here is the longer version of the description: The Ada embedding model deployment name within the same Azure OpenAI resource. When this parameter is provided, Azure OpenAI on your data will use an internal call to evaluate the Ada embedding model, rather than calling the Azure OpenAI endpoint. This enables you to use vector search when the Azure OpenAI resource is in private networks.

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_vector_search.tsp:
```diff
@@ -0,0 +1,63 @@
+  @doc("Azure OpenAI Embedding deployment name.")
+  deploymentName: "DeploymentName",
+
+  @doc("The embedding model id defined in the search service. Currently only supported by Elasticsearch.")
```
nit: "id" --> "ID" (here and elsewhere)

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_vector_search.tsp:
```diff
@@ -0,0 +1,63 @@
+  @doc("The embedding dependency based on model id")
+  type: OnYourDataEmbeddingDependencyType.modelId;
+
+  @doc("The embedding model id build inside the search service. Currently only supported by Elasticsearch.")
```
nit: Align with the description of the enum value: "The embedding model ID defined in the search service. Currently only supported by Elasticsearch."

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_vector_search.tsp:
```diff
@@ -0,0 +1,63 @@
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("The embedding dependency based on endpoint URL")
+model OnYourDataEmbeddingEndpointDependency
```
nit: It could be valuable to describe what an "embedding dependency" is.

wangyuantao replied:
Embedding dependency tells the service how to calculate the embedding of the intent to retrieve the data using vector search. I just look at how Azure AI Search name it - they call it vectorizer, seems to be a better name than embedding dependency, thoughts? https://learn.microsoft.com/en-us/azure/search/vector-search-how-to-configure-vectorizer#define-a-vectorizer

joseharriaga commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_common_options.tsp:
```diff
@@ -0,0 +1,30 @@
+alias OnYourDataCommonChatExtensionConfiguration = {
+  @doc("The authentication option to access the data.")
+  @projectedName("csharp", "Authentication")
+  authentication?: OnYourDataAuthenticationOptions;
```
Should this property be required? 🤔

wangyuantao replied:
We use "system_assigned_managed_identity" as default authentication, if not provided explicitly. Then the request can be simpler and curl friendly. Good point - we should document it clearly what is the default authentication.

trrwilson replied:
Updating to this; wordy, but hopefully helpful:
```
  @doc("""
  The authentication method to use when accessing the defined data source.
  Each data source type supports a specific set of available authentication methods; please see the documentation of
  the data source for supported mechanisms.
  If not otherwise provided, On Your Data will attempt to use System Managed Identity (default credential)
  authentication.
  """)
```

jpalvarezl committed:
Annotated Key as removed

jpalvarezl committed:
Restored LRO as private methods

trrwilson committed:
Merge branch 'main' into jpalvarezl/aoai_extensions_reorg

trrwilson committed:
Merge branch 'main' into user/travisw/aoai-2023-12-01-preview-typespec

trrwilson committed:
Merge branch 'jpalvarezl/aoai_extensions_reorg' into user/travisw/aoai-2023-12-01-preview-typespec

trrwilson committed:
PR feedback

wangyuantao commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cosmosdb_options.tsp:
```diff
@@ -0,0 +1,53 @@
+  @projectedName("csharp", "ContainerName")
+  containerName: string;
+
+  @doc("The index name name of Azure Cosmos DB.")
```
We are currently supporting "Azure Cosmos DB for MongoDB vCore" only. It has index name: https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search#use-the-createindexes-template-to-create-a-vector-index.

It's a good idea to clarify in the description we only support "Azure Cosmos DB for MongoDB vCore", not any of "Azure Cosmos DB". For the discriminator enum value, not sure if "azure_cosmosdb_for_mongodb_vcore" will be too long?

trrwilson replied:
For now, I'll add clarification to the comments:
```
Parameters to use when configuring Azure OpenAI On Your Data chat extensions when using Azure Cosmos DB for
MongoDB vCore.
```
```
@doc("The MongoDB vCore database name to use with Azure Cosmos DB.")
```
```
@doc("The MongoDB vCore index name to use with Azure Cosmos DB.")
```

jpalvarezl committed:
Hid LROs (#26898)

trrwilson committed:
Merge branch 'main' into user/travisw/aoai-2023-12-01-preview-typespec

trrwilson committed:
format, recompile, backport optional filename for whisper

trrwilson committed:
Merge branch 'user/travisw/aoai-2023-12-01-preview-typespec' of https://github.com/Azure/azure-rest-api-specs into user/travisw/aoai-2023-12-01-preview-typespec

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cosmosdb_options.tsp:
```diff
@@ -0,0 +1,53 @@
+
+  @doc("Customized field mapping behavior to use when interacting with the search index.")
+  @projectedName("csharp", "FieldMappingOptions")
+  fieldsMapping: AzureCosmosDBFieldMappingOptions;
```
Per earlier with my ludicrous "hot dog eating instructions" example, I think "field mapping options" works here.

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions_onyourdata_azure_cosmosdb_options.tsp:
```diff
@@ -0,0 +1,53 @@
+  @projectedName("csharp", "DatabaseName")
+  databaseName: string;
+
+  @doc("The container name name of Azure Cosmos DB.")
```
Agreed; updating to "The name of the Azure Cosmos DB resource container." Likewise for "...name of the Pinecone database index."

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions/oyd/azure_machinelearning_index_options.tsp:
```diff
@@ -0,0 +1,47 @@
+  @projectedName("csharp", "Version")
+  version: string;
+
+  @doc("Search filter. Only supported if the MLIndex is of type AzureSearch.")
```
Expanded.

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions/oyd/azure_machinelearning_index_options.tsp:
```diff
@@ -0,0 +1,47 @@
+}
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("Parameters for the Azure Machine Learning index chat extension.")
```
Replaced all docs instances in this AML file with "vector index." Left types as-is for now.

trrwilson commented on specification/cognitiveservices/OpenAI.Inference/models/completions/extensions/oyd/pinecone_options.tsp:
```diff
@@ -0,0 +1,82 @@
+  @projectedName("csharp", "Environment")
+  environment: string;
+
+  @doc("The index name name of Pinecone.")
```
Good catch; like the CosmosDB one, replaced this with "The name of the Pinecone database index."

trrwilson committed:
PR feedback: naming/docs updates

mssfang committed:
Added TypeSpec customization support for Java (#26916)

* Update spell checking instructions (#26874)

Having one custom-words.txt for the entire repo has been causing a number of merging issues so we are trying to move away from it and instead have teams use override in the cspell configuration file directly.

* Update readme.python.md (#26894)

* Update readme.python.md

* Update readme.python.md

* Update readme.python.md

* Update readme.python.md

---------

Co-authored-by: Yuchao Yan <yuchaoyan@microsoft.com>

* [Cognitive Services - Azure AI Content Safety] Update package print name. (#26911)

* Update package print name.

* update.

* update name.

* Added TypeSpec customization support for Java

---------

Co-authored-by: Wes Haggard <weshaggard@users.noreply.github.com>
Co-authored-by: ChenxiJiang333 <119990644+ChenxiJiang333@users.noreply.github.com>
Co-authored-by: Yuchao Yan <yuchaoyan@microsoft.com>
Co-authored-by: jiezhou_microsoft <94604058+JieZhou000@users.noreply.github.com>

jpalvarezl committed:
Renamed customization class (#26919)

trrwilson committed:
Update specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp

Co-authored-by: Deyaaeldeen Almahallawi <dealmaha@microsoft.com>

trrwilson committed:
Merge branch 'main' into user/travisw/aoai-2023-12-01-preview-typespec

trrwilson committed:
recompile after accepted/committed feedback fix

deyaaeldeen commented on specification/cognitiveservices/OpenAI.Inference/models/completions/azure_content_filtering.tsp:
```diff
@@ -0,0 +1,175 @@
+
+  @doc("Content filtering results for this prompt")
+  @projectedName("json", "content_filter_results")
+  contentFilterResults?: ContentFilterResultDetailsForPrompt;
```
Should this be required? I wonder why would we get an entry with a prompt index but no content filter results.Should this be required? I wonder why would we get an entry with a prompt index but no content filter results.

trrwilson replied:
Good question. Reviewing the .json document, [it doesnlt look like EITHER of the two properties are marked as required](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.json#L1152), which is particularly strange to imagine. I'll follow up to see if there's a reason for this but I'm inclined to agree with your assessment and make these both required until proven otherwise; it seems like it's just a spec omission that we can correct.

deyaaeldeen commented on specification/cognitiveservices/OpenAI.Inference/models/completions/chat_completions.tsp:
```diff
@@ -3,75 +3,37 @@ import "@typespec/http";
-    64 characters.
-    """)
-  @added(ServiceApiVersions.v2023_07_01_Preview)
-  name?: string;
```
OpenAI is adding `name` back to their spec: https://github.com/openai/openai-openapi/issues/118

trrwilson replied:
Good eye! I believe we have this matched correctly: System, User, and Assistant messages have the optional "name" property while "Tool" and "Function" do not. That appears to be how the fix landed in their OpenAPI3 YAML file, too.

trrwilson committed:
PR feedback: non-optional content_filter_results on prompt model

trrwilson committed:
very minor change: one file fix via npx prettier

trrwilson committed:
missing second half of prettier fix: copied file to data-plane

deyaaeldeen commented on specification/cognitiveservices/OpenAI.Inference/models/completions/tools.tsp:
```diff
@@ -0,0 +1,97 @@
+  type: "function";
+
+  @doc("The name of the function that should be called.")
+  name: "string";
```
The type should be string and not the literal "string", I think.
```suggestion
  name: string;
```

deyaaeldeen replied:
I opened https://github.com/Azure/azure-rest-api-specs/pull/26960.
